tool/mapeditor.re

```diff
@@ -50,60 +65,102 @@ static MAP_EDITOR_COMPONENT = Component {
+                let map_list = Tas::list_maps();
+                enter_ui(Ui::new_filechooser("Map to play", map_list, fn(input: string) {
+                    MAP_EDITOR_STATE.map_name = input;
+                    if map_list.contains(input) {
+                        MAP_EDITOR_STATE.map = Tas::load_map(input);
+                        Tas::apply_map(MAP_EDITOR_STATE.map);
+                    } else {
+                        MAP_EDITOR_STATE.map = Tas::current_map();
+                    };
+                    MAP_EDITOR_EDIT_LABEL.text = "Stop Map Editor";
+                    add_component(MAP_EDITOR_COMPONENT);
+                    MAP_EDITOR_STATE.mode = MapEditorMode::Play;
+                    MOVEMENT_STATE.enable_fly = false;
+                    leave_ui();
+                    leave_ui();
+                    leave_ui();
+                }));
+            }
+        },
+    });
+
+    let delete_map_button =  UiElement::Button(UiButton {
+        label: Text { text: "Delete Map" },
+        onclick: fn(label: Text) {
+            fn create_map_editor_delete_map_menu() -> Ui {
+                let map_list = Tas::list_maps();
+                Ui::new_filechooser("Map to delete", map_list, fn(input: string) {
+                    if map_list.contains(input) {
+                        Tas::remove_map(input);
                         leave_ui();
-                    }));
-                }
-            },
-        }),
-        UiElement::Button(UiButton {
-            label: Text { text: "Delete Map" },
-            onclick: fn(label: Text) {
-                fn create_map_editor_delete_map_menu() -> Ui {
-                    let map_list = Tas::list_maps();
-                    Ui::new_filechooser("Map to delete", map_list, fn(input: string) {
-                        if map_list.contains(input) {
-                            Tas::remove_map(input);
-                            leave_ui();
-                            enter_ui(create_map_editor_delete_map_menu());
-                        }
-                    })
-                }
-                enter_ui(create_map_editor_delete_map_menu());
-            },
-        }),
-        UiElement::Button(UiButton {
-            label: Text { text: "Back" },
-            onclick: fn(label: Text) { leave_ui() },
-        }),
-    ))
+                        enter_ui(create_map_editor_delete_map_menu());
+                    }
+                })
+            }
+            enter_ui(create_map_editor_delete_map_menu());
+        },
+    });
+
+    let back_button = UiElement::Button(UiButton {
+       label: Text { text: "Back" },
+       onclick: fn(label: Text) { leave_ui() },
+    });
+
+    if MAP_EDITOR_STATE.mode == MapEditorMode::None {
+        return Ui::new("Map Editor", List::of(edit_map_button, play_map_button, delete_map_button, back_button))
+    } else {
+        return Ui::new("Map Editor", List::of(edit_map_button, delete_map_button, back_button))
+    }
```

Comment:
Instead of
```rust
let foo = ...;
let bar = ...;
Ui::new(..., List::of(foo, maybe(bar)));
```
you can do
```rust
let mut list = List::new();
list.push(...);
if maybe() { list.push(...); }
Ui::new(..., list);
```